src/traversals/allsimplepaths.jl

```diff
@@ -0,0 +1,188 @@
+using DataStructures
+
+"""
+    all_simple_paths(g, source, targets, cutoff)
+
+Returns an iterator that generates all simple paths in the graph `g` from `source` to `targets`.
+If `cutoff` is given, the paths' lengths are limited to equal or less than `cutoff`.
+Note that the length of a path is defined as the number of edges, not the number of elements.
+ex. the path length of `[1, 2, 3]` is two.
+Internally, a DFS algorithm is used to search paths.
+
+# Examples
+
+```jldoctest
+julia> using LightGraphs
+julia> g = complete_graph(4)
+julia> collect(all_simple_paths(g, 1, [4]))
+5-element Array{Array{Int64,1},1}:
+ [1, 4]
+ [1, 3, 4]
+ [1, 3, 2, 4]
+ [1, 2, 4]
+ [1, 2, 3, 4]
+ ```
+"""
+function all_simple_paths(g::AbstractGraph, source::T, targets::Vector{T}; cutoff::T=typemax(T)) where T <: Integer
+    return SimplePathIterator(g, source, Set(targets), cutoff=cutoff)
+end
+
+
+"""
+    all_simple_paths(g, source, target, cutoff)
+
+This function is equivalent to `all_simple_paths(g, source, [target], cutoff)`.
+This is provided for convenience.
+
+See also `all_simple_paths(g, source, targets, cutoff)`.
+"""
+function all_simple_paths(g::AbstractGraph, source::T, target::T; cutoff::T=typemax(T)) where T <: Integer
+    return SimplePathIterator(g, source, Set(target), cutoff=cutoff)
+end
+
+
+"""
+    SimplePathIterator{T <: Integer}
+
+Iterator that generates all simple paths.
+The iterator holds the condition specified in `all_simple_path` function.
+"""
+struct SimplePathIterator{T <: Integer}
+    g::AbstractGraph
+    source::T  # Starting node
+    targets::Set{T}  # Target nodes
+    cutoff::T  # Max length of resulting paths
+
+    function SimplePathIterator(g::AbstractGraph, source::T, targets::Set{T}; cutoff::T=typemax(T)) where T <: Integer
+        new{T}(g, source, targets, cutoff)
+    end
+end
+
+
+"""
+    SimplePathIteratorState{T <: Integer}
+
+SimplePathIterator's state.
+"""
+mutable struct SimplePathIteratorState{T <: Integer}
+    stack::Stack{Vector{T}}  # Store information used to restore iteration of child nodes. Each vector has two elements which are a parent node and an index of children.
+    visited::Stack{T}  # Store current path candidate
+    queued_targets::Vector{T}  # Store rest targets if path length reached cutoff.
+    function SimplePathIteratorState(spi::SimplePathIterator{T}) where T <: Integer
+        stack = Stack{Vector{T}}()
+        visited = Stack{T}()
+        queued_targets = Vector{T}()
+        push!(visited, spi.source)  # Add a starting node to the path candidate
+        push!(stack, [spi.source, 1])  # Add a child node with index = 1
+        new{T}(stack, visited, queued_targets)
+    end
+end
+
+"""
+    function _stepback!(state)
+
+A helper function that updates iterator state.
+For internal use only.
```

Comment:
```suggestion
A helper function that updates the simple path iterator state.
```
Internal use is already implied by the underscore